lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1416,53 +1416,111 @@ static EnableOp getLastEnableOp(SeqOp parent) {
   return nullptr;
 }
 
-/// Removes common tail enable operations for sequential 'then'/'else'
-/// branches inside an 'if' operation.
-///
-///   if %a with %A {                       if %a with %A {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   else {                          ->    } else {
-///     seq { ... calyx.enable @B }           seq { ... }
-///   }                                     }
-///                                         calyx.enable @B
-static LogicalResult eliminateCommonTailEnable(IfOp ifOp,
-                                               PatternRewriter &rewriter) {
-  // Check if the branches exist.
+/// Returns a mapping of {Enabled Group, EnableOp} for all EnableOps within the
+/// immediate ParOp's body.
+static llvm::StringMap<EnableOp> getAllEnableOpsInImmediateBody(ParOp parent) {
+  llvm::StringMap<EnableOp> enables;
+  Block *body = parent.getBody();
+  for (EnableOp op : body->getOps<EnableOp>())
+    enables.insert(std::pair(op.groupName(), op));
+
+  return enables;
+}
+
+/// Removes common tail enable operations for then and else regions encompassed
+/// in SeqOp and ParOp. This canonicalization is stringent about not entering
+/// nested control operations, as this may cause unintentional changes in
+/// behavior.
+static LogicalResult eliminateCommonTail(IfOp ifOp, PatternRewriter &rewriter) {
   if (!ifOp.thenRegionExists() || !ifOp.elseRegionExists())
     return failure();
+  rewriter.setInsertionPointAfter(ifOp);
 
-  auto &thenOpStructureOp = ifOp.getThenBody()->front();
-  auto &elseOpStructureOp = ifOp.getElseBody()->front();
-  // TODO(circt/#1861): ParOps have less restrictive conditions.
-  if (isa<ParOp>(thenOpStructureOp) || isa<ParOp>(elseOpStructureOp))
-    return failure();
-
-  // At this point, only sequential operations are valid inside the branches.
-  auto thenSeqOp = dyn_cast<SeqOp>(thenOpStructureOp);
-  auto elseSeqOp = dyn_cast<SeqOp>(elseOpStructureOp);
-  assert(thenSeqOp && elseSeqOp &&
-         "expected nested seq ops in both branches of a calyx.IfOp");
-
-  EnableOp lastThenEnableOp = getLastEnableOp(thenSeqOp);
-  EnableOp lastElseEnableOp = getLastEnableOp(elseSeqOp);
+  auto &thenControl = ifOp.getThenBody()->front();
+  auto &elseControl = ifOp.getElseBody()->front();
+  if (auto r1 = dyn_cast<ParOp>(thenControl);
+      auto r2 = dyn_cast<ParOp>(elseControl)) {
+    ///   if %a with @G {                  if %a with @G {
+    ///     par {                            par { ... }
+    ///       ...                          } else {
+    ///       calyx.enable @A     ->         par { ... }
+    ///       calyx.enable @B              }
+    ///     }                              calyx.enable @A
+    ///   else {                           calyx.enable @B
```

Comment:
What is the context that `calyx.enable @A, calyx.enable @B` gets extracted to? Are we expecting the `if` to be nested inside a `seq`, and `calyx.enable @A, calyx.enable @B` to be placed within a `par` following the `if``?

Reply:
If the `then`, `else` regions are nested in `seq`s, then we should nest the `if` and extracted `enable`s in a `seq` as well. Similar for `par`. This ensures we are still maintaining similar code behavior. In the worst case, we have `seq { seq { ... } }`, but this gets canonicalized away. Did that answer your question?

Reply:
I understand; my main concern was just that it wasn't included in the code example, i.e. groups going from clearly being within a `par` group to be extracted into the parent group, which is not shown.

Reply:
What do you mean by "parent group"?

The example doesn't show the `IfOp`'s parent mostly because there are two different cases, and the behavior is somewhat expected, e.g. if an `EnableOp` is driven sequentially, it will still be driven sequentially when pulled out.

If you think this isn't clear, I can add the parent to the example.